### DISCUSSION TITLE
Searching for component like Vue-autonumeric

### DISCUSSION THREAD

**Discussion by AdagioB** [quasarframework/quasar#12301] [General - Components / Directives / etc] [answered]
Hi

Note that I'm using UMD and can't use those scripts to install components. As far as I know, I have to link directly to js and vue files

I'm trying to update from Vue 2 to 3 and Quasar from 1.xx to 2.xx, but I have one issue I haven't been able to find a solution for. 
My application is using a lot of numeric input fields, where the users most of the times needs to write decimal numbers. The issue is that my application is only used by people in Denmark, where we use comma as decimal separator. Quasars input only works with the US way of writing decimal numbers as far as I can tell. I tried using masks, but that just made the user experience worse. I also have a requirement that the field can both be empty (null value) or 0. I remember before I found Vue-autonumeric, there were some components that couldn't tell the difference between 0 and null

My thirst thought was to get a newer version of Vue-autonumeric, but from what I can see, there is no version that works with Vue 3. I did somehow (don't remember how) get a hold of a js file, where I can add it to my application and it correctly shows up. Unfortunately changing the value does not do anything. The value sent to the component stays the same

Someone in another thread posted a link to [Vue-currencry-input](https://dm4t2.github.io/vue-currency-input/guide.html), where the js file is located [here](https://unpkg.com/vue-currency-input@2.4.0/dist/index.cjs.js)
I tried to set up a simple test using their examples, but Vue comes with this warning: 'Failed to resolve component: currency-input'
The exampled can be found [on codepen](https://codepen.io/Adagio_B/pen/wvPGxpj?editors=1011). My guess is that this component might be my best hope. They even have an example of this working with Quasar, but all their examples doesn't work in UMD as far as I can tell

In the Quasar Input, there is some mentioning of v-money directive and money component, but when I'm trying them in my application. They just come with errors like "Failed to resolve directive: money"

Does anybody know what I'm doing wrong or if there is a better component out there that works?

**Comment by mickey-bimlive** [comment, marked as answer]
I have the same challenge, as a heavy user of VueAutoNumeric which doesn't support Vue 3. So let me share my solution with VueCurrencyInput, https://www.npmjs.com/package/vue-currency-input

I succeeded in doing a sniff test (still on Vue 2) of my Quasar app with VueCurrencyInput. My app is a PWA though, not UMD.

I derived my VueCurrencyInput test code from https://codesandbox.io/s/using-vue-currency-input-with-quasar-veevalidate-gokw6?from-embed=&file=/src/App.vue:35-295

My experience with VMoney for decimal input is very bad. It has some basic flaws. I posted that in the old Quasar forum a while ago.

**Reply by mickey-bimlive** [reply]
One thing I remember from my test is that I had to put the code for the component as a new component "dezimal-eingabe.vue" into my exsting \components folder.  The code for the component follows the sample from
https://dm4t2.github.io/vue-currency-input/guide.html#creating-a-custom-component (I'll post it separately, below).

It has (as my choice for naming):
`name: "QDecimalInput"`

In an existing Quasar component of my app, I imported that code for that new component, 
```
import QDecimalInput from "../components/dezimal-eingabe.vue";
...
components: QDecimalInput
```
In the template part of my Quasar component, I used that component as follows:
`<QDecimalInput v-model="xyz" :options="{ ... }"  />`

The content in "{ ... }" follows the VueCurrencyInput docs, in my case, I tried:
```
:options="{
  currency: 'EUR', // Ist erforderlich, wird aber unterdrückt
  currencyDisplay: 'hidden', // 'EUR' nicht anzeigen
  autoDecimalDigits: false, // Whether the decimal symbol is inserted automatically, using the last inputted digits as decimal digits. Default is false (the decimal symbol needs to be inserted manually).
  precision: 2, // 2 Nachkommastellen
  // distractionFree: false, // So nicht dokumentiert in API Reference
  hideCurrencySymbolOnFocus: true, // Whether to hide the currency symbol on focus. Default is true.
  hideGroupingSeparatorOnFocus: true, // Whether to hide the grouping separator on focus. Default is true.
  hideNegligibleDecimalDigitsOnFocus: true, // Whether to hide negligible decimal digits on focus. Default is true.
  valueRange: { min: -1000, max: 1000 }, // The range of accepted values as object {min, max}. Default is undefined (no value range).
  // The validation is triggered on blur and automatically sets the respective threshold if out of range.
  useGrouping: true // Trennzeichen für Tausenderstellen, '.' mit deutschem Locale
}"
```

**Comment by mickey-bimlive** [comment]
Here is my code that I used for the component (dezimal-eingabe.vue) - sorry some of the comments are in German...
I extended the VueCurrencyInput input field with two buttons (per v-slots) to allow the user to increment and decrement the decimal value in the field.

```
<!-- Komponente für ein Input-Feld mit Dezimalzahleingabe -->
<!-- Vorlagen: -->
<!-- https://dm4t2.github.io/vue-currency-input/guide.html#creating-a-custom-component -->
<!-- https://codesandbox.io/s/using-vue-currency-input-with-quasar-veevalidate-gokw6?from-embed=&file=/src/QCurrencyInput.vue -->
<template>
  <!--  -->
  <div>
    <q-input
      ref="inputRef"
      :value="formattedValue"
      dense
      square
      bg-color="teal-2"
      standout="bg-teal-4 text-black"
      input-class="text-black text-right"
      label="[m über NHN]"
      hint="NHN-Niveau mit VueCurrencyInput"
    >
      <template v-slot:prepend>
        <q-btn
          dense
          @click="myIncrementOrDecrement(true)"
          icon="arrow_drop_up"
          color="grey-4"
          text-color="black"
          style="width:20px"
        >
        </q-btn>
      </template>
      <template v-slot:append>
        <q-btn
          dense
          @click="myIncrementOrDecrement(false)"
          icon="arrow_drop_down"
          color="grey-4"
          text-color="black"
          style="width:20px"
        >
        </q-btn>
      </template>
    </q-input>
    <br />
    <br />
    Aktueller Wert im VModel: <code>{{ $props.value }}</code>
    <br />
    Typ des VModels: <code>{{ typeof $props.value }}</code>
    <br />
    Durch VueCurrencyInput formatierter Wert: <code>{{ formattedValue }}</code>
    <br />
    Typ des formatierten Werts: <code>{{ typeof formattedValue }}</code>
    <br />
    VueCurrencyInput Props: <code>{{ $props }}</code>
  </div>
</template>
//***======================================================================================
<script>
import { watch } from '@vue/composition-api'; // Für watch() unten NOTWENDIG!
import { parse, useCurrencyInput } from "vue-currency-input";

//--------------------------------------------------------------------------------------------------------------------------------
export default {
  // name: 'ComponentName',
  name: "QDecimalInput", // Für Dezimalzahleingabe gewählter Komponentenname
  props: {
    value: Number, // Vue 2: value; Vue 3: modelValue
    options: Object
  },
  setup(props) {
    const { inputRef, formattedValue, setOptions, setValue } = useCurrencyInput(props.options)
    // ACHTUNG: Hier watch() eingefügt nach Vorlage auf https://dm4t2.github.io/vue-currency-input/guide.html#external-props-changes
    // Ist notwendig, wenn das VModel nicht nur durch Benutzereingabe, sondern durch externe Funktionen (z.B. Inkrement/Dekrement) geändert wird!
    //==================================================================================
    watch(() => props.value, (value) => { // Vue 2: props.value; Vue 3: props.modelValue
      setValue(value)
    })
    //==================================================================================
    watch(() => props.options, (options) => {
      setOptions(options)
    })
    return { inputRef, formattedValue, setValue };
  },
  //-----------------------------------------------------------------------------------------------------------------------------------------------------
  data() {
    return {};
  },
  //-----------------------------------------------------------------------------------------------------------------------------------------------------
  methods: {
    myIncrementOrDecrement(incrementNotDecrement = true) {
      // Zugriff auf das VModel per obiger Prop value:
      // console.log("$props.value: ", this.$props.value);
      console.log("value: ", this.value) // ACHTUNG: this.value ist äquivalent zu this.$props.value!
      console.log(
        "typeof value: ",
        typeof this.value
      );
      // Hier sind keine Typkonvertierungen notwendig wie mit VueAutoNumeric!!!
      // Das VModel hat bereits den Type number
      let number = this.value;
      if (incrementNotDecrement) {
        number = number + 0.01
      } else {
        number = number - 0.01
      }
      number = Math.round(number * 100) / 100; // Präventive Rundungskorrektur, siehe
      // https://stackoverflow.com/questions/38561833/obtain-two-decimal-places-in-javascript-without-rounding-to-the-next-bigger-numb
      // Hier auch KEINE Rückkonvertierung mit Helper-Funktion nach Dezimalstring durchgeführt
      //=====================================================================
      // Emittieren eines Standard-Input-Events - dieses wird standardmäßig vom VModel verarbeitet, siehe
      // https://stackoverflow.com/questions/54782010/using-v-bind-and-v-on-instead-of-v-model-vuejs 
      // Dies funktioniert allerdings nur in Verbindung mit obigem watch() - vermutlich reagiert die Komponente selbst
      // auf diesen @input-Event und aktualisiert dann das VModel.
      //----------------------------------------------------------------------------------------------------------------------
      this.$emit("input", number) // Dies bewirkt eine Aktualisierung des VModels
    },
  }
};
</script>
```

**Reply by AdagioB** [reply]
Thanks, I'll take a look at it as soon as my works-schedule gives me time to work on this again. Don't worry about the German comments, I can understand it perfectly fine as I have been living near the German borders most of my life 👍
